design-proposals/klip-11-DQL.md

```diff
@@ -1 +1,745 @@
+In particular, we should not implement (cf. "Design" section below):
+
+* re-introduce `CREATE TABLE` statement to create RDBMS style tables that allow insert/update/delete statements
+* allow for complex queries over table state, e.g., aggregation and joins
+* we might not need to implement the new `AS OF <timestamp>` clause in the first iteration
+* we might not need to support terminating queries over streams right away
+
+We limit this KLIP to not introduce too much syntactic sugar to express queries.
+This may make some queries unnecessary "complex/clumsy" but the goal is to illustrate to underlying principles of the language.
+We can add syntactic sugar after we agree on the basic design of the language.
+
+
+
+## Value/Return
+
+The "Dual Query Language" should be easier to understand and easier to use.
+It will allow to support nested queries seamlessly and unifies query semantics (persistent vs. transient queries).
+Furthermore, the language is designed to be "future proof", i.e., it considers potential future features and thus allows
+to extend the capabilities of KSQL without the need to change the language in a breaking way.
+
+
+
+## Public APIS
+
+We propose the following language changes:
+
+* introduce `TABLE(<stream>)` operator: takes a STREAM as input, and changes its semantic interpretation from facts to updates; it returns a TABLE
+  * note: the returned TABLE is not an RDBMS style table, but rather a MV
+  * note: it's not specified if the returned TABLE is materialized into a state store or represented as a changelog
+    (it's up the the KSQL optimizer or KS runtime to make a materialization decision)
+  * the `TABLE()` operator consumes it's input STREAM "from beginning"
+  * we could extend `TABLE()` to also accept a TABLE as input if it makes the language smoother;
+    it's obvious that `TABLE(<table>)` would be idempotent, i.e., `TABLE(t) = t`
+* introduce `STREAM(<table>)` operator: takes a TABLE as input, and returns a STREAM that contains a record for each update to the TABLE
+  * note: the returned STREAM is a fact stream, and would be processed with corresponding STREAM semantics (it's semantically not a changelog stream)
+  * note: the input TABLE can either be an RDBMS style table of a MV
+  * by default, the output STREAM contains one record for each input table record (full table scan on the table generation when the query was issues)
+    plus a record for each input table update
+  * we could extend `STREAM()` to also accept a STREAM as input if it makes the language smoother;
+  it's obvious that `STREAM(<stream>)` would be idempotent, i.e., `STREAM(s) = s`
+* Introduce an `AS OF <timestamp>` clause for RDBMS style queries against tables
+* Introduce `CREATE MATERIALIZED VIEW` statement
+* Deprecate `CREATE TABLE` statement (in favor of the new `CREATE MATERIALIZED VIEW` statement) and eventually remove it
+* Deprecate `EMIT CHANGES` clause
```

Comment:
> Furthermore, removing this may cause frustration as we've already decided to do this and users are learning to leverage this clause.

Well. That's is the basic problem of a fast moving project -- you break stuff... I agree it's not ideal, but atm I believe it's the right thing to do... If there is a good technical argument to keep `EMIT CHANGES`, I can be happily convinced, what bring me to your question about "output rate control":

This KLIP does not cover "output rate control" atm, and I agree that it is an important aspect that we should embed in the language. If we believe that it's important enough to cover right now, I can extend the KLIP to cover it (maybe even bringing `EMIT CHANGES` or something similar back into the picture).

The main argument of this KLIP itself stay valid with or without rate control: `EMIT CHANGES` should not be required to get an output stream: a "stream query" returns an output stream naturally and requiring `EMIT CHANGES` to, eg, filter an input stream, seems undesirable. Similar for getting the changelog stream from a table, `TABLE(<stream>)` is proposed. Hence, I consider both aspect semi-orthogonal to each other.

Thoughts?

Reply:
Yeah I think we're in agreement here 😄

Reply:
> EDIT: somewhat in agreement - I think EMIT on a table query should turn it into a subscription (push) query

Where do you see the disagreement? The KLIP suggests to replace `EMIT` with `STREAM(<table>)` what implies that it would be a "stream query" (aka push query). Seems to be the same thing?

Reply:
Let's discuss this on https://github.com/confluentinc/ksql/pull/3799#discussion_r349285809